src/System.Text.Primitives/System/Text/Encoders/Utf8Validator.cs

```diff
@@ -0,0 +1,252 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Diagnostics;
+
+namespace System.Buffers.Text
+{
+    /// <summary>
+    /// Allows processing discontiguous buffers of UTF-8 data, letting the caller know whether
+    /// any data seen so far is invalid UTF-8. The caller should invoke <see cref="TryConsume(ReadOnlySpan{byte, bool})"/>
+    /// in a loop until all data has been consumed.
+    /// </summary>
+    /// <remarks>
+    /// This type is a mutable struct.
+    /// </remarks>
+    [DebuggerDisplay("{DebuggerDisplay,nq}")]
+    public struct Utf8Validator
+    {
+        /// <summary>
+        /// A sequence that represents a fresh no-partial-buffer validity checker.
+        /// </summary>
+        private const uint DefaultSequence = 0;
+
+        /// <summary>
+        /// A sequence that represents a validity checker which has seen invalid data.
+        /// </summary>
+        private const uint InvalidSequence = unchecked((uint)~0);
```

Comment:
Why not just uint.MaxValue?

Reply:
`~0` says "give me an integer where every bit is set" more clearly than `uint.MaxValue` does.